src/test_fixture/mod.rs

```diff
@@ -52,17 +54,94 @@ pub struct MaliciousContext<'a, F: Field> {
-        let acc = v.accumulator();
-
-        MaliciousContext {
-            ctx: ctx.upgrade_to_malicious(acc, v.r_share().clone()),
-            validator: v,
-        }
-    })
+///
+/// # Panics
+/// If there are any errors upgrading the inputs to `MaliciousReplicated` it will panic
+/// This can happen if there are any problems with the multiplications used to do so.
+pub async fn make_malicious_contexts<'a, F: Field, R: RngCore>(
+    test_world: &'a TestWorld,
+    inputs: &[F],
+    rng: &mut R,
+) -> (
+    [ProtocolContext<'a, MaliciousReplicated<F>, F>; 3],
+    [SecurityValidator<F>; 3],
+    [Vec<MaliciousReplicated<F>>; 3],
+)
+where
+    Standard: Distribution<F>,
+{
+    let contexts = make_contexts(test_world);
+    let validators: Vec<_> = contexts
+        .iter()
+        .map(|ctx| SecurityValidator::new(ctx.narrow("MaliciousValidate")))
+        .collect();
+
+    let mut helper0_shares = Vec::with_capacity(inputs.len());
+    let mut helper1_shares = Vec::with_capacity(inputs.len());
+    let mut helper2_shares = Vec::with_capacity(inputs.len());
+    for input in inputs {
+        let [sh0, sh1, sh2] = share(*input, rng);
+        helper0_shares.push(sh0);
+        helper1_shares.push(sh1);
+        helper2_shares.push(sh2);
+    }
+
+    let malicious_stuff = try_join_all(
+        zip(
+            zip(contexts.into_iter(), validators.iter()),
+            [helper0_shares, helper1_shares, helper2_shares],
+        )
+        .map(|((ctx, v), shares)| async move {
+            try_join_all(
+                zip(
+                    repeat(v.r_share()),
+                    zip(repeat(v.accumulator().clone()), zip(shares, repeat(ctx))),
+                )
+                .enumerate()
+                .map(|(i, (r_share, (acc, (s, ctx))))| async move {
+                    let record_id = RecordId::from(i);
+                    ctx.narrow("upgrade_inputs")
+                        .upgrade_to_malicious(acc, r_share.clone(), record_id, s)
+                        .await
+                }),
+            )
+            .await
+        }),
+    )
+    .await
+    .unwrap();
+
+    let malicious_inputs = [
+        malicious_stuff[0].iter().map(|(_, x)| x.clone()).collect(),
+        malicious_stuff[1].iter().map(|(_, x)| x.clone()).collect(),
+        malicious_stuff[2].iter().map(|(_, x)| x.clone()).collect(),
+    ];
+
+    let malicious_contexts = [
+        malicious_stuff[0][0].0.clone(),
+        malicious_stuff[1][0].0.clone(),
+        malicious_stuff[2][0].0.clone(),
+    ];
+
+    (
+        malicious_contexts,
+        validators.try_into().unwrap(),
+        malicious_inputs,
+    )
+}
+
+/// # Errors
+/// If any of the calls to validate fail, this will throw an error
+pub async fn validate_circuit<F: Field>(
+    contexts: [ProtocolContext<'_, MaliciousReplicated<F>, F>; 3],
+    validators: [SecurityValidator<F>; 3],
+) -> Result<(), Error> {
+    let _validation_results = try_join_all(
```

Comment:
You shouldn't need to assign this to anything, it's a type of `()`, right?

Reply:
I think try_join_all returns a vector, so it's a Vec<Result<(), Error>>

Reply:
Then maybe you should unwrap() it rather than throwing it away.